File: JavascriptAlgorithm/Tree/minDepth-111.js
var minDepth1 = function (root) {
  if (!root) {
    return 0;
  }
  let depth = Infinity;
  const dfs = (level, node) => {
    if (!node) {
      return;
    }
    if (!node.left && !node.right) {
      depth = Math.min(level, depth);
    }
    dfs(level + 1, node.left);
    dfs(level + 1, node.right);
  };
  dfs(1, root);
  return depth;
};

var minDepth = function (node) {
  if (!node) {
    return Infinity;
  }
  if (!node.left && !node.right) {
    return 1;
  }
  return Math.min(minDepth(node.left) + 1, minDepth(node.right) + 1);
};
var minDepth2 = function (root) {
  return !root ? 0 : minDepth(root);
};
export { minDepth1, minDepth2 };

File: JavaAlgorithm/src/main/java/algo/tzashinorpu/ThirdRound/Chapter07/minDepth_111.java
package algo.tzashinorpu.ThirdRound.Chapter07;

import algo.tzashinorpu.ThirdRound.TreeNode;

public class minDepth_111 {
    // 最小深度是从根节点到最近叶子节点的最短路径上的节点数量
    private int depth = Integer.MAX_VALUE;

    public int minDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        dfs(root, 0);
        return depth;
    }

    private void dfs(TreeNode root, int level) {
        if (root.left == null && root.right == null) {
            depth = Math.min(depth, level);
            return;
        }
        if (root.left != null) dfs(root.left, level + 1);
        if (root.right != null) dfs(root.right, level + 1);
    }
}
